src/main/java/com/digitalpebble/storm/crawler/bolt/SiteMapParserBolt.java

```diff
@@ -0,0 +1,178 @@
+import com.digitalpebble.storm.crawler.protocol.HttpHeaders;
+import com.digitalpebble.storm.crawler.util.KeyValues;
+
+import crawlercommons.sitemaps.AbstractSiteMap;
+import crawlercommons.sitemaps.SiteMap;
+import crawlercommons.sitemaps.SiteMapIndex;
+import crawlercommons.sitemaps.SiteMapURL;
+import crawlercommons.sitemaps.SiteMapURL.ChangeFrequency;
+
+/**
+ * Extracts URLs from sitemap files. The parsing is triggered by the presence of
+ * 'isSitemap=true' in the metadata. Any tuple which does not have this
+ * key/value in the metadata is simply passed on to the default stream, whereas
+ * any URLs extracted from the sitemaps is sent to the 'status' field.
+ **/
+public class SiteMapParserBolt extends BaseRichBolt {
+
+    private OutputCollector collector;
+    public static final String isSitemapKey = "isSitemap";
+
+    private boolean strictMode = false;
+
+    private static final org.slf4j.Logger LOG = LoggerFactory
+            .getLogger(SiteMapParserBolt.class);
+
+    @Override
+    public void execute(Tuple tuple) {
+        HashMap<String, String[]> metadata = (HashMap<String, String[]>) tuple
+                .getValueByField("metadata");
+
+        // TODO check that we have the right number of fields ?
+        String isBoolean = KeyValues.getValue(isSitemapKey, metadata);
+        if (StringUtils.isBlank(isBoolean)
+                || !isBoolean.equalsIgnoreCase("true")) {
+            // just pass it on
+            this.collector.emit(tuple.getValues());
+            this.collector.ack(tuple);
+            return;
+        }
+
+        // it does have the right key/value
+        byte[] content = tuple.getBinaryByField("content");
+        String url = tuple.getStringByField("url");
+        String ct = KeyValues.getValue(HttpHeaders.CONTENT_TYPE, metadata);
+        List<Values> outlinks = parseSiteMap(url, content, ct);
+        if (outlinks == null) {
+            // likely to happen ad lib - just ack
+            // error has been logged
+            this.collector.ack(tuple);
+            return;
+        }
+
+        // send to status stream
+        for (Values ol : outlinks) {
+            collector.emit(Constants.StatusStreamName, new Values(ol));
+        }
+
+        // marking the main URL as successfully fetched
+        collector.emit(Constants.StatusStreamName, new Values(url, metadata,
+                Status.FETCHED));
+
+        this.collector.ack(tuple);
+    }
+
+    private List<Values> parseSiteMap(String url, byte[] content,
+            String contentType) {
+
+        crawlercommons.sitemaps.SiteMapParser parser = new crawlercommons.sitemaps.SiteMapParser(
+                strictMode);
+
+        AbstractSiteMap siteMap = null;
+        try {
+            siteMap = parser.parseSiteMap(contentType, content, new URL(url));
+        } catch (Exception e) {
+            LOG.error("Exception while parsing sitemap", e);
+            return null;
```

Comment:
even more so now that we also send a tuple on the status stream for the main URL. Fixed